core/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -63,19 +69,66 @@
         if (canRecover(indexShard)) {
             if (indexShard.routingEntry().restoreSource() != null) {
                 throw new IllegalStateException("can't recover - restore source is not null");
             }
             return executeRecovery(indexShard, () -> {
                 logger.debug("starting recovery from store ...");
-                internalRecoverFromStore(indexShard, indexShouldExists);
+                internalRecoverFromStore(indexShard, indexShouldExists, () -> {});
+            });
+        }
+        return false;
+    }
+
+    boolean recoverFromLocalShards(Client client, final IndexShard indexShard, final List<LocalShardSnapshot> shards) throws IOException {
+        if (canRecover(indexShard)) {
+            if (indexShard.routingEntry().restoreSource() != null) {
+                throw new IllegalStateException("can't recover - restore source is not null");
+            }
+            if (shards.isEmpty()) {
+                throw new IllegalArgumentException("shards must not be empty");
+            }
+            Set<Index> indices = shards.stream().map((s) -> s.index).collect(Collectors.toSet());
+            if (indices.size() > 1) {
+                throw new IllegalArgumentException("can't add shards from more than one index");
+            }
+            IndexMetaData metaData = shards.get(0).shard.indexSettings().getIndexMetaData();
+            for (ObjectObjectCursor<String, MappingMetaData> mapping : metaData.getMappings()) {
+                client.admin().indices().preparePutMapping(indexShard.shardId().getIndexName())
+                    .setType(mapping.key)
+                    .setSource(mapping.value.source().string())
+                    .get();
+            }
+            for (ObjectObjectCursor<String, MappingMetaData> mapping : metaData.getMappings()) {
+                indexShard.mapperService().merge(mapping.key,mapping.value.source(), MapperService.MergeReason.MAPPING_RECOVERY, true);
+            }
+            return executeRecovery(indexShard, () -> {
+                logger.debug("starting recovery from local shards {}", shards);
+                internalRecoverFromStore(indexShard, false, () -> {
+                    try {
+                        final Engine engine = indexShard.getEngine();
+                        // nocommit - do we need stats for this?
+                        engine.addIndices(shards.stream().map(s->s.getSnapshotDirectory())
+                            .collect(Collectors.toList()).toArray(new Directory[0]));
+                        engine.flush(true, true); // commit the added indices
+                        Store store = indexShard.store();
+                        SegmentInfos si = store.readLastCommittedSegmentsInfo();
+                        final RecoveryState.Index index = indexShard.recoveryState().getIndex();
+                        final Directory directory = store.directory();
+                        for (String name : Lucene.files(si)) {
```

Comment:
This code can be shared by the two recoveries (from shards and local), types no?